### PR TITLE
Kdesktop 707 automatically restart the app in case of crash

### PR DESCRIPTION
**Pull request by herve-er**
Implementation of self restart after a server or client crash with protection over crash/restart loop.

### CODE REVIEW

File: src/gui/appclient.cpp [hunk 1]
```diff
@@ -501,6 +489,21 @@ void AppClient::onQuit() {
     quit();
 }
 
+void AppClient::onServerDisconnected() {
+#if NDEBUG
+    if(connectToServer()){
+    		qCInfo(lcAppClient) << "Reconnected to server";
+    }else {
```
Comment (luc-guyot-infomaniak) [inline src/gui/appclient.cpp:496]: Do you need some help to automate C++ formatting on your IDE?

File: src/gui/appclient.cpp [hunk 2]
```diff
@@ -581,6 +584,76 @@ void AppClient::setupLogging() {
...
+    // Connect to server
+    int count = 0;
+    while (!CommClient::instance()->connectToServer(_commPort) && count < CONNECTION_TRIALS) {
+        qCCritical(lcAppClient()) << "Connection to server failed!";
+        count++;
+    }
+
+    if (count == CONNECTION_TRIALS) {
+        QMessageBox msgBox;
+        msgBox.setText(tr("Unable to connect to application server!"));
```
Comment (luc-guyot-infomaniak) [inline src/gui/appclient.cpp:624]: Idem

File: src/gui/appclient.cpp [hunk 3]
```diff
@@ -581,6 +584,76 @@ void AppClient::setupLogging() {
...
+            qCWarning(lcAppClient()) << "Check of comm status failed";
+            count++;
+        } else {
+            break;
+        }
+    }
+
+    if (count == CHECKCOMMSTATUS_TRIALS) {
+        QMessageBox msgBox;
+        msgBox.setText(tr("The application server did not respond on time!"));
```
Comment (luc-guyot-infomaniak) [inline src/gui/appclient.cpp:648]: Idem

File: src/gui/appclient.cpp [hunk 4]
```diff
@@ -501,6 +489,21 @@ void AppClient::onQuit() {
...
 }
 
+void AppClient::onServerDisconnected() {
+#if NDEBUG
+    if(connectToServer()){
+    		qCInfo(lcAppClient) << "Reconnected to server";
+    }else {
+			qCCritical(lcAppClient) << "Failed to reconnect to server";
+            //TODO: Implement a limit of reconnection trials per x minutes
+			startServerAndKillClient();
```
Comment (luc-guyot-infomaniak) [inline src/gui/appclient.cpp:499]: What about `startServerAndDie`?

File: src/gui/appclient.cpp [hunk 5]
```diff
@@ -69,14 +70,32 @@ static const QList<QString> fontFiles = QList<QString>() << QString(":/client/re
...
+
 AppClient::AppClient(int &argc, char **argv)
     : SharedTools::QtSingleApplication(Theme::instance()->appClientName(), argc, argv),
       _gui(nullptr),
       _theme(Theme::instance()),
       _logExpire(0),
       _logFlush(false),
       _logDebug(false),
-      _debugMode(false) {
+      _debugMode(false),
```
Comment (luc-guyot-infomaniak) [inline src/gui/appclient.cpp:97]: Please check the advice of Sonar: https://sonarcloud.io/project/issues?resolved=false&pullRequest=37&id=Infomaniak_desktop-kDrive&open=AY6piGSrMvoBiF_fbDAK.

File: src/gui/appclient.cpp [hunk 6]
```diff
@@ -581,6 +584,76 @@ void AppClient::setupLogging() {
...
+    clientProcess->startDetached();
+
+    QTimer::singleShot(0, qApp, SLOT(quit()));
+}
+
+bool AppClient::connectToServer() {
+    // Check if a commPort is provided
+    if (_commPort == 0) {
+        QMessageBox msgBox;
+        msgBox.setText(tr("No comm port provided! Clean restart in progress..."));
```
Comment (luc-guyot-infomaniak) [inline src/gui/appclient.cpp:608]: What does "clean" mean in this context?

File: src/gui/appclient.cpp (hunk 4 above)
Comment (herve-er) [reply src/gui/appclient.cpp:499]: Thanks you for the suggestion, updated in last commit.

File: src/libparms/db/parmsdb.cpp [hunk 7]
```diff
@@ -3175,4 +3291,94 @@ bool ParmsDb::selectAllMigrationSelectiveSync(std::vector<MigrationSelectiveSync
...
+}
+
+
+bool ParmsDb::updateLastServerSelfRestartTime(int64_t lastServertRestartTime) {
+    const std::lock_guard<std::mutex> lock(_mutex);
+
+    if (lastServertRestartTime == -1) {
+        auto now = std::chrono::system_clock::now();
+        auto now_s = std::chrono::time_point_cast<std::chrono::seconds>(now);
+        auto value = now_s.time_since_epoch();
```
Comment (luc-guyot-infomaniak) [inline src/libparms/db/parmsdb.cpp:3339]:
```suggestion
       const auto value = now_s.time_since_epoch();
```

File: src/libparms/db/parmsdb.cpp [hunk 8]
```diff
@@ -3175,4 +3291,94 @@ bool ParmsDb::selectAllMigrationSelectiveSync(std::vector<MigrationSelectiveSync
...
+
+    return true;
+}
+
+bool ParmsDb::updateLastClientSelfRestartTime(int64_t lastClientRestartTime) {
+    const std::lock_guard<std::mutex> lock(_mutex);
+
+
+    if (lastClientRestartTime == -1) {
+        auto now = std::chrono::system_clock::now();
```
Comment (luc-guyot-infomaniak) [inline src/libparms/db/parmsdb.cpp:3363]: You can use `const` for each of these declarations.

File: src/libparms/db/parmsdb.cpp [hunk 9]
```diff
@@ -3175,4 +3291,94 @@ bool ParmsDb::selectAllMigrationSelectiveSync(std::vector<MigrationSelectiveSync
...
+        auto now_s = std::chrono::time_point_cast<std::chrono::seconds>(now);
+        auto value = now_s.time_since_epoch();
+        lastClientRestartTime = value.count();
+    }
+
+    ASSERT(queryResetAndClearBindings(UPDATE_SELF_RESTARTER_CLIENT_REQUEST_ID));
+    ASSERT(queryBindValue(UPDATE_SELF_RESTARTER_CLIENT_REQUEST_ID, 1, lastClientRestartTime));
+
+    int errId;
```
Comment (luc-guyot-infomaniak) [inline src/libparms/db/parmsdb.cpp:3372]:
```suggestion
    int errId = 0;
```

File: src/server/appserver.cpp [hunk 10]
```diff
@@ -2755,6 +2796,34 @@ void AppServer::setupProxy() {
...
+		return true;
+	}
+}
+
+bool AppServer::clientCrashedRecently(int second) {
+    int64_t now_s = std::chrono::time_point_cast<std::chrono::seconds>(std::chrono::system_clock::now()).time_since_epoch().count();
+
+    int64_t lastClientCrash;
+    KDC::ParmsDb::instance()->selectLastClientSelfRestartTime(lastClientCrash);
+    auto diff = now_s - lastClientCrash;
```
Comment (luc-guyot-infomaniak) [inline src/server/appserver.cpp:2818]:
```suggestion
   const auto diff = nowSecond - lastClientCrash;
```

File: src/server/appserver.cpp [hunk 11]
```diff
@@ -2755,6 +2796,34 @@ void AppServer::setupProxy() {
...
+    if (diff > second) {
+		return false;
+	} else {
+        LOG_WARN(_logger, "Server crashed recently: " << diff << " seconds ago");
+		return true;
+	}
+}
+
+bool AppServer::clientCrashedRecently(int second) {
+    int64_t now_s = std::chrono::time_point_cast<std::chrono::seconds>(std::chrono::system_clock::now()).time_since_epoch().count();
```
Comment (luc-guyot-infomaniak) [inline src/server/appserver.cpp:2814]:
```suggestion
   const int64_t nowSeconds = std::chrono::time_point_cast<std::chrono::seconds>(std::chrono::system_clock::now()).time_since_epoch().count();
```

File: src/server/appserver.cpp [hunk 12]
```diff
@@ -2755,6 +2796,34 @@ void AppServer::setupProxy() {
     Proxy::instance(ParametersCache::instance()->parameters().proxyConfig());
 }
 
+bool AppServer::serverCrashedRecently(int second) {
+    int64_t now_s = std::chrono::time_point_cast<std::chrono::seconds>(std::chrono::system_clock::now()).time_since_epoch().count();
+
+    int64_t lastServerCrash; 
+    KDC::ParmsDb::instance()->selectLastServerSelfRestartTime(lastServerCrash);
+    auto diff = now_s - lastServerCrash;
```
Comment (luc-guyot-infomaniak) [inline src/server/appserver.cpp:2804]:
```suggestion
    const auto diff = nowSeconds - lastServerCrash;
```

File: src/server/appserver.cpp [hunk 13]
```diff
@@ -2755,6 +2796,34 @@ void AppServer::setupProxy() {
     Proxy::instance(ParametersCache::instance()->parameters().proxyConfig());
 }
 
+bool AppServer::serverCrashedRecently(int second) {
+    int64_t now_s = std::chrono::time_point_cast<std::chrono::seconds>(std::chrono::system_clock::now()).time_since_epoch().count();
```
Comment (luc-guyot-infomaniak) [inline src/server/appserver.cpp:2800]:
```suggestion
    const int64_t nowSeconds = std::chrono::time_point_cast<std::chrono::seconds>(std::chrono::system_clock::now()).time_since_epoch().count();
```

File: src/server/appserver.cpp [hunk 14]
```diff
@@ -2755,6 +2796,34 @@ void AppServer::setupProxy() {
...
 
+bool AppServer::serverCrashedRecently(int second) {
+    const int64_t nowSeconds = std::chrono::time_point_cast<std::chrono::seconds>(std::chrono::system_clock::now()).time_since_epoch().count();
+
+    int64_t lastServerCrash; 
+    KDC::ParmsDb::instance()->selectLastServerSelfRestartTime(lastServerCrash);
+    const auto diff = nowSeconds - lastServerCrash;
+    if (diff > second) {
+		return false;
+	} else {
```
Comment (luc-guyot-infomaniak) [inline src/server/appserver.cpp:2807]: The indentation looks odd.

File: src/server/appserver.h [hunk 15]
```diff
@@ -142,6 +143,8 @@ class AppServer : public SharedTools::QtSingleApplication {
         void parseOptions(const QStringList &);
         void initLogging() noexcept(false);
         void setupProxy();
+        bool serverCrashedRecently(int second = 60 /*Allow one server self restart per minute (default)*/);
```
Comment (luc-guyot-infomaniak) [inline src/server/appserver.h:146]:
```suggestion
        bool serverCrashedRecently(int seconds = 60 /*Allow one server self restart per minute (default)*/);
```

File: src/libparms/db/parmsdb.cpp [hunk 16]
```diff
@@ -535,6 +564,32 @@ ParmsDb::ParmsDb(const std::filesystem::path &dbPath, const std::string &version
...
+    std::string error;
+
+    ASSERT(queryResetAndClearBindings(INSERT_SELF_RESTARTER_REQUEST_ID));
+    ASSERT(queryBindValue(INSERT_SELF_RESTARTER_REQUEST_ID, 1, 0));
+    ASSERT(queryBindValue(INSERT_SELF_RESTARTER_REQUEST_ID, 2, 0));
+
+    if (!queryExec(INSERT_SELF_RESTARTER_REQUEST_ID, errId, error)) {
+		LOG_WARN(_logger, "Error running query: " << INSERT_SELF_RESTARTER_REQUEST_ID);
+		return false;
+	}
```
[the remark this answers is not in the document]
Comment (herve-er) [reply src/libparms/db/parmsdb.cpp:586]: I did, strangely it appear only on git, not in my IDE or in codespace. Will investigate this.

File: src/server/appserver.cpp (hunk 14 above)
Comment (herve-er) [reply src/server/appserver.cpp:2807]: Done in next commits.

File: src/libparms/db/parmsdb.cpp [hunk 17]
```diff
@@ -3175,4 +3291,94 @@ bool ParmsDb::selectAllMigrationSelectiveSync(std::vector<MigrationSelectiveSync
     return true;
 }
 
+bool ParmsDb::selectLastServerSelfRestartTime(int64_t &lastServerRestartTime) {
+    const std::lock_guard<std::mutex> lock(_mutex);
+
+	ASSERT(queryResetAndClearBindings(SELECT_SELF_RESTARTER_REQUEST_ID));
```
[the remark this answers is not in the document]
Comment (herve-er) [reply src/libparms/db/parmsdb.cpp]: Done in next commits.

File: src/libparms/db/parmsdb.cpp [hunk 18]
```diff
@@ -1329,6 +1440,11 @@ bool ParmsDb::initData() {
         return false;
     }
 
+    if (!insertDefaultSelfRestarterData()) {
+		LOG_WARN(_logger, "Error in insertDefaultSelfRestarterData");
```
[the remark this answers is not in the document]
Comment (herve-er) [reply src/libparms/db/parmsdb.cpp]: Done in next commits.

File: src/libparms/db/parmsdb.cpp [hunk 19]
```diff
@@ -1227,12 +1295,36 @@ bool ParmsDb::prepare() {
...
+    if (!queryPrepare(UPDATE_SELF_RESTARTER_CLIENT_REQUEST_ID, UPDATE_SELF_RESTARTER_CLIENT_REQUEST, false, errId, error)) {
+        queryFree(UPDATE_SELF_RESTARTER_CLIENT_REQUEST_ID);
+        return sqlFail(UPDATE_SELF_RESTARTER_CLIENT_REQUEST_ID, error);
+    }
+
+    ASSERT(queryCreate(INSERT_SELF_RESTARTER_REQUEST_ID));
+    if (!queryPrepare(INSERT_SELF_RESTARTER_REQUEST_ID, INSERT_SELF_RESTARTER_REQUEST, false, errId, error)) {
+		queryFree(INSERT_SELF_RESTARTER_REQUEST_ID);
+		return sqlFail(INSERT_SELF_RESTARTER_REQUEST_ID, error);
+	}
```
[the remark this answers is not in the document]
Comment (herve-er) [reply src/libparms/db/parmsdb.cpp]: Done in next commits.

File: src/libparms/db/parmsdb.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (herve-er) [reply src/libparms/db/parmsdb.cpp]: Done in next commits.

File: src/gui/appclient.cpp [hunk 20]
```diff
@@ -124,44 +135,15 @@ AppClient::AppClient(int &argc, char **argv)
...
-        std::cout << "Connection to server failed!" << std::endl;
-        count++;
-    }
-
-    if (count == CONNECTION_TRIALS) {
-        QMessageBox msgBox;
-        msgBox.setText(tr("Unable to connect to application server!"));
-        msgBox.exec();
-        QTimer::singleShot(0, qApp, SLOT(quit()));
+    if(connectToServer()){
```
[the remark this answers is not in the document]
Comment (herve-er) [reply src/gui/appclient.cpp]: Done in next commits.

**Comment by ClementKunz** [comment]
You should try to fix as many as possible issues detected by Sonar inorder for our code quality to improve

**Comment by herve-er** [comment]
@ClementKunz I have done some of the recommended changes,  but some are not applicable (ie: new QProcess).

File: src/gui/appclient.cpp [hunk 21]
```diff
@@ -501,6 +483,17 @@ void AppClient::onQuit() {
     quit();
 }
 
+void AppClient::onServerDisconnected() {
+#if NDEBUG
+    qCCritical(lcAppClient) << "Server disconnected | Closing client";
```
[the remark this answers is not in the document]
Comment (luc-guyot-infomaniak) [reply src/gui/appclient.cpp:485]: I would put `static const auto msg = tr("The server got disconnected. Closing client");` at the top of the function and reuse it twice.

File: src/server/appserver.cpp [hunk 22]
```diff
@@ -2755,6 +2796,36 @@ void AppServer::setupProxy() {
...
+    const auto diff = nowSeconds - lastServerCrash;
+    if (diff > second) {
+        return false;
+    } else {
+        LOG_WARN(_logger, "Server crashed recently: " << diff << " seconds ago");
+        return true;
+    }
+}
+
+bool AppServer::clientCrashedRecently(int second) {
```
Comment (luc-guyot-infomaniak) [inline src/server/appserver.cpp:2814]:
```suggestion
bool AppServer::clientCrashedRecently(int seconds) {
```

File: src/gui/appclient.cpp (hunk 21 above)
[the remark this answers is not in the document]
Comment (herve-er) [reply src/gui/appclient.cpp]: Done, and add log even in release.

File: src/server/appserver.cpp [hunk 23]
```diff
@@ -2755,6 +2796,36 @@ void AppServer::setupProxy() {
...
+        LOG_WARN(_logger, "Server crashed recently: " << diff << " seconds ago");
+        return true;
+    }
+}
+
+bool AppServer::clientCrashedRecently(int seconds) {
+    const int64_t nowSeconds =
+        std::chrono::time_point_cast<std::chrono::seconds>(std::chrono::system_clock::now()).time_since_epoch().count();
+
+    int64_t lastClientCrash;
```
Comment (ClementKunz) [inline src/server/appserver.cpp:2818]: Initialization

Comment (herve-er) [reply src/server/appserver.cpp]: Corrected in commit 08c8427.

File: src/libparms/db/parmsdb.cpp [hunk 24]
```diff
@@ -3175,4 +3288,87 @@ bool ParmsDb::selectAllMigrationSelectiveSync(std::vector<MigrationSelectiveSync
     return true;
 }
 
+bool ParmsDb::selectLastServerSelfRestartTime(int64_t &lastServerRestartTime) {
+    const std::scoped_lock lock(_mutex);
+
+    ASSERT(queryResetAndClearBindings(SELECT_SELF_RESTARTER_REQUEST_ID));
+    bool found;
```
Comment (ClementKunz) [inline src/libparms/db/parmsdb.cpp:3295]: Still missing initialization

**Review by ClementKunz — APPROVED** [review]

**Pull request merged**